agent/consul/xdscapacity/capacity.go

```diff
@@ -0,0 +1,203 @@
+
+import (
+	"context"
+	"math"
+
+	"github.com/armon/go-metrics"
+	"github.com/armon/go-metrics/prometheus"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-memdb"
+	"golang.org/x/time/rate"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/proto/pbpeering"
+)
+
+var StatsGauges = []prometheus.GaugeDefinition{
+	{
+		Name: []string{"xds", "server", "idealStreamsMax"},
+		Help: "The maximum number of xDS streams per server, chosen to achieve a roughly even spread of load across servers.",
+	},
+}
+
+// errorMargin is amount to which we allow a server to be over-occupied,
+// expressed as a percentage (between 0 and 1).
+//
+// We allow 10% more than the ideal number of streams per server.
+const errorMargin = 0.1
+
+// Controller determines the ideal number of xDS streams for the server to
+// handle and enforces it using the given SessionLimiter.
+//
+// We aim for a roughly even spread of streams between servers in the cluster
+// and, to that end, limit the number of streams each server can handle to:
+//
+//	(<number of proxies> / <number of healthy servers>) + <error margin>
+//
+// Controller receives changes to the number of healthy servers from the
+// autopilot delegate. It queries the state store's catalog tables to discover
+// the number of registered proxy (sidecar and gateway) services.
+type Controller struct {
+	cfg Config
+
+	serverCh chan uint32
+	doneCh   chan struct{}
+
+	prevMaxSessions uint32
+	prevRateLimit   rate.Limit
+}
+
+// Config contains the dependencies for Controller.
+type Config struct {
+	Logger         hclog.Logger
+	GetStore       func() Store
+	SessionLimiter SessionLimiter
+}
+
+// SessionLimiter is used to enforce the session limit to achieve the ideal
+// spread of xDS streams between servers.
+type SessionLimiter interface {
+	SetMaxSessions(maxSessions uint32)
+	SetDrainRateLimit(rateLimit rate.Limit)
+}
+
+// NewController creates a new capacity controller with the given config.
+//
+// Call Run to start the control-loop.
+func NewController(cfg Config) *Controller {
+	return &Controller{
+		cfg:      cfg,
+		serverCh: make(chan uint32),
+		doneCh:   make(chan struct{}),
+	}
+}
+
+// Run the control-loop until the given context is canceled or reaches its
+// deadline.
+func (a *Controller) Run(ctx context.Context) {
+	defer close(a.doneCh)
+
+	ws, numProxies, err := a.countProxies()
+	if err != nil {
+		a.cfg.Logger.Error("failed to count proxy services", "error", err)
+	}
+
+	var numServers uint32
+	for {
+		select {
+		case s := <-a.serverCh:
+			numServers = s
+			a.updateMaxSessions(numServers, numProxies)
+		case <-ws.WatchCh(ctx):
+			var count uint32
+			ws, count, err = a.countProxies()
+			if err == nil {
+				numProxies = count
+				a.updateDrainRateLimit(numProxies)
+				a.updateMaxSessions(numServers, numProxies)
+			} else {
+				a.cfg.Logger.Error("failed to count proxy services", "error", err)
```

Comment:
❗ (blocking)

Is there any feasible way to recover here? I think if we hit this case we will never again update our proxy count. Could we do something with exponential backoff where we infinitely retry until counting proxies returns a non-error result?